[Bug] [core] in Partition table case, call IndexBootstrap.bootstrapType function throw exception

### Search before asking

- [X] I searched in the [issues](https://github.com/apache/paimon/issues) and found nothing similar.


### Paimon version

0.9-SNAPSHOT

### Compute Engine

java api

### Minimal reproduce step

none

### What doesn't meet your expectations?

call IndexBootstrap.bootstrapType function throw exception, in Partition table case. 

```java
Schema.Builder schemaBuilder = Schema.newBuilder();
        schemaBuilder.column("id", DataTypes.INT().notNull(), "pk");
        schemaBuilder.column("name", DataTypes.STRING(), "name");
        schemaBuilder.column("dt", DataTypes.DATE().notNull(), "date");
        schemaBuilder.primaryKey("id","dt");
        schemaBuilder.partitionKeys("dt");
        schemaBuilder.option("bucket", "5");
        Schema schema = schemaBuilder.build();
```
execption msg:
```java
Exception in thread "main" java.lang.RuntimeException: java.lang.IllegalArgumentException: Field names must be unique. Found duplicates: [dt]
	at com.hawk.paimon.file.Main.main(Main.java:35)
Caused by: java.lang.IllegalArgumentException: Field names must be unique. Found duplicates: [dt]
	at org.apache.paimon.types.RowType.validateFields(RowType.java:183)
	at org.apache.paimon.types.RowType.<init>(RowType.java:66)
	at org.apache.paimon.types.RowType.<init>(RowType.java:70)
	at org.apache.paimon.schema.TableSchema.projectedLogicalRowType(TableSchema.java:272)
	at org.apache.paimon.crosspartition.IndexBootstrap.bootstrapType(IndexBootstrap.java:143)
	at com.hawk.paimon.file.Main.main(Main.java:29)
```
### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

--- a/paimon-core/src/main/java/org/apache/paimon/crosspartition/IndexBootstrap.java
+++ b/paimon-core/src/main/java/org/apache/paimon/crosspartition/IndexBootstrap.java
@@ -142,6 +142,7 @@ public class IndexBootstrap implements Serializable {
                 new ArrayList<>(
                         schema.projectedLogicalRowType(
                                         Stream.concat(primaryKeys.stream(), partitionKeys.stream())
+                                                .distinct()
                                                 .collect(Collectors.toList()))
                                 .getFields());
         bootstrapFields.add(

--- a/paimon-core/src/test/java/org/apache/paimon/crosspartition/IndexBootstrapTest.java
+++ b/paimon-core/src/test/java/org/apache/paimon/crosspartition/IndexBootstrapTest.java
@@ -117,8 +117,9 @@ public class IndexBootstrapTest extends TableTestBase {
     public void testBootstrapType() throws Exception {
         FileStoreTable table = (FileStoreTable) createTable();
         RowType indexRowType = IndexBootstrap.bootstrapType(table.schema());
+        assertThat(indexRowType.getFieldNames()).contains(BUCKET_FIELD);
         assertThat(indexRowType.getFieldIndex(BUCKET_FIELD))
-                .isEqualTo(RowType.currentHighestFieldId(indexRowType.getFields()));
+                .isEqualTo(2); // the last field is bucket, which is not in table schema
     }
 
     @Test

--- a/paimon-core/src/test/java/org/apache/paimon/crosspartition/IndexBootstrapTest.java
+++ b/paimon-core/src/test/java/org/apache/paimon/crosspartition/IndexBootstrapTest.java
@@ -27,10 +27,12 @@ import org.apache.paimon.io.DataFileMeta;
 import org.apache.paimon.manifest.FileSource;
 import org.apache.paimon.options.Options;
 import org.apache.paimon.schema.Schema;
+import org.apache.paimon.table.FileStoreTable;
 import org.apache.paimon.table.Table;
 import org.apache.paimon.table.TableTestBase;
 import org.apache.paimon.table.source.DataSplit;
 import org.apache.paimon.types.DataTypes;
+import org.apache.paimon.types.RowType;
 import org.apache.paimon.utils.Pair;
 
 import org.junit.jupiter.api.Test;
@@ -43,6 +45,7 @@ import java.util.Collections;
 import java.util.List;
 import java.util.function.Consumer;
 
+import static org.apache.paimon.crosspartition.IndexBootstrap.BUCKET_FIELD;
 import static org.apache.paimon.crosspartition.IndexBootstrap.filterSplit;
 import static org.apache.paimon.data.BinaryRow.EMPTY_ROW;
 import static org.apache.paimon.stats.SimpleStats.EMPTY_STATS;
@@ -53,20 +56,7 @@ public class IndexBootstrapTest extends TableTestBase {
 
     @Test
     public void testBoostrap() throws Exception {
-        Identifier identifier = identifier("T");
-        Options options = new Options();
-        options.set(CoreOptions.BUCKET, -1);
-        Schema schema =
-                Schema.newBuilder()
-                        .column("pt", DataTypes.INT())
-                        .column("col", DataTypes.INT())
-                        .column("pk", DataTypes.INT())
-                        .primaryKey("pk")
-                        .partitionKeys("pt")
-                        .options(options.toMap())
-                        .build();
-        catalog.createTable(identifier, schema, true);
-        Table table = catalog.getTable(identifier);
+        Table table = createTable();
 
         write(
                 table,
@@ -104,6 +94,31 @@ public class IndexBootstrapTest extends TableTestBase {
         // here, (this is good, beneficial for query speed) but TableTestBase.after will check leak
         // streams. So sleep here to avoid unstable.
         Thread.sleep(1000);
+    }
+
+    private Table createTable() throws Exception {
+        Identifier identifier = identifier("T");
+        Options options = new Options();
+        options.set(CoreOptions.BUCKET, -1);
+        Schema schema =
+                Schema.newBuilder()
+                        .column("pt", DataTypes.INT())
+                        .column("col", DataTypes.INT())
+                        .column("pk", DataTypes.INT())
+                        .primaryKey("pk")
+                        .partitionKeys("pt")
+                        .options(options.toMap())
+                        .build();
+        catalog.createTable(identifier, schema, true);
+        return catalog.getTable(identifier);
+    }
+
+    @Test
+    public void testBootstrapType() throws Exception {
+        FileStoreTable table = (FileStoreTable) createTable();
+        RowType indexRowType = IndexBootstrap.bootstrapType(table.schema());
+        assertThat(indexRowType.getFieldIndex(BUCKET_FIELD))
+                .isEqualTo(RowType.currentHighestFieldId(indexRowType.getFields()));
     }
 
     @Test

--- a/paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/GlobalDynamicBucketTableITCase.java
+++ b/paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/GlobalDynamicBucketTableITCase.java
@@ -174,23 +174,23 @@ public class GlobalDynamicBucketTableITCase extends CatalogITCaseBase {
     @Test
     public void testBootstrapType() throws Exception {
         Catalog catalog = CatalogFactory.createCatalog(CatalogContext.create(new Path(path)));
-        FileStoreTable T = (FileStoreTable) catalog.getTable(Identifier.create("default", "T"));
-        FileStoreTable partial_update_t =
+        FileStoreTable t = (FileStoreTable) catalog.getTable(Identifier.create("default", "T"));
+        FileStoreTable partialUpdateT =
                 (FileStoreTable) catalog.getTable(Identifier.create("default", "partial_update_t"));
-        FileStoreTable first_row_t =
+        FileStoreTable firstRowT =
                 (FileStoreTable) catalog.getTable(Identifier.create("default", "first_row_t"));
-        assertThat(IndexBootstrap.bootstrapType(T.schema()).getFieldNames()).contains(BUCKET_FIELD);
-        assertThat(IndexBootstrap.bootstrapType(partial_update_t.schema()).getFieldNames())
+        assertThat(IndexBootstrap.bootstrapType(t.schema()).getFieldNames()).contains(BUCKET_FIELD);
+        assertThat(IndexBootstrap.bootstrapType(partialUpdateT.schema()).getFieldNames())
                 .contains(BUCKET_FIELD);
-        assertThat(IndexBootstrap.bootstrapType(first_row_t.schema()).getFieldNames())
+        assertThat(IndexBootstrap.bootstrapType(firstRowT.schema()).getFieldNames())
                 .contains(BUCKET_FIELD);
-        assertThat(IndexBootstrap.bootstrapType(T.schema()).getFieldIndex(BUCKET_FIELD))
+        assertThat(IndexBootstrap.bootstrapType(t.schema()).getFieldIndex(BUCKET_FIELD))
                 .isEqualTo(2);
         assertThat(
-                        IndexBootstrap.bootstrapType(partial_update_t.schema())
+                        IndexBootstrap.bootstrapType(partialUpdateT.schema())
                                 .getFieldIndex(BUCKET_FIELD))
                 .isEqualTo(2);
-        assertThat(IndexBootstrap.bootstrapType(first_row_t.schema()).getFieldIndex(BUCKET_FIELD))
+        assertThat(IndexBootstrap.bootstrapType(firstRowT.schema()).getFieldIndex(BUCKET_FIELD))
                 .isEqualTo(2);
     }
 }

--- a/paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/GlobalDynamicBucketTableITCase.java
+++ b/paimon-flink/paimon-flink-common/src/test/java/org/apache/paimon/flink/GlobalDynamicBucketTableITCase.java
@@ -18,12 +18,21 @@
 
 package org.apache.paimon.flink;
 
+import org.apache.paimon.catalog.Catalog;
+import org.apache.paimon.catalog.CatalogContext;
+import org.apache.paimon.catalog.CatalogFactory;
+import org.apache.paimon.catalog.Identifier;
+import org.apache.paimon.crosspartition.IndexBootstrap;
+import org.apache.paimon.fs.Path;
+import org.apache.paimon.table.FileStoreTable;
+
 import org.apache.flink.types.Row;
 import org.junit.jupiter.api.Test;
 
 import java.util.Arrays;
 import java.util.List;
 
+import static org.apache.paimon.crosspartition.IndexBootstrap.BUCKET_FIELD;
 import static org.assertj.core.api.Assertions.assertThat;
 
 /** ITCase for batch file store. */
@@ -160,5 +169,28 @@ public class GlobalDynamicBucketTableITCase extends CatalogITCaseBase {
         sql("insert into large_t select * from src");
         sql("insert into large_t select * from src");
         assertThat(sql("select k, count(*) from large_t group by k having count(*) > 1")).isEmpty();
+    }
+
+    @Test
+    public void testBootstrapType() throws Exception {
+        Catalog catalog = CatalogFactory.createCatalog(CatalogContext.create(new Path(path)));
+        FileStoreTable T = (FileStoreTable) catalog.getTable(Identifier.create("default", "T"));
+        FileStoreTable partial_update_t =
+                (FileStoreTable) catalog.getTable(Identifier.create("default", "partial_update_t"));
+        FileStoreTable first_row_t =
+                (FileStoreTable) catalog.getTable(Identifier.create("default", "first_row_t"));
+        assertThat(IndexBootstrap.bootstrapType(T.schema()).getFieldNames()).contains(BUCKET_FIELD);
+        assertThat(IndexBootstrap.bootstrapType(partial_update_t.schema()).getFieldNames())
+                .contains(BUCKET_FIELD);
+        assertThat(IndexBootstrap.bootstrapType(first_row_t.schema()).getFieldNames())
+                .contains(BUCKET_FIELD);
+        assertThat(IndexBootstrap.bootstrapType(T.schema()).getFieldIndex(BUCKET_FIELD))
+                .isEqualTo(2);
+        assertThat(
+                        IndexBootstrap.bootstrapType(partial_update_t.schema())
+                                .getFieldIndex(BUCKET_FIELD))
+                .isEqualTo(2);
+        assertThat(IndexBootstrap.bootstrapType(first_row_t.schema()).getFieldIndex(BUCKET_FIELD))
+                .isEqualTo(2);
     }
 }

